File: solana-monitoring/fetch/get_token_transfers.py
import argparse
import json
import requests
from collections import defaultdict
from time import sleep

# === CLI Args ===
parser = argparse.ArgumentParser(description="Parse Solana token transfer types from the signatures and extract direction flows by token transfer type.")
parser.add_argument("signature_file", help="Path to the JSON file with getBlock results")
parser.add_argument("--output", default="direction_counts_by_token_transfer_type.json", help="Output JSON file")
args = parser.parse_args()

SIGNATURE_FILE = args.signature_file
OUTPUT_FILE = args.output
RPC_URL = "https://devnet.helius-rpc.com/?api-key=934c08b5-0c44-4549-bda8-30587b901642"

# === Load and parse JSON signatures ===
with open(SIGNATURE_FILE, 'r') as f:
    raw_data = json.load(f)

signatures = []
for entry in raw_data:
    result = entry.get("result", {})
    sigs = result.get("signatures", [])
    signatures.extend(sigs)

print(f"📦 Loaded {len(signatures)} signatures.")

# === Process each transaction ===
direction_counts = defaultdict(int)
headers = { "Content-Type": "application/json" }

token_transfers = []

for i, sig in enumerate(signatures):
    try:
        print(f"[{i+1}/{len(signatures)}] Fetching token transfers for {sig}...")
        payload = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "getTokenTransfers",
            "params": [sig, "jsonParsed"]
        }

        response = requests.post(RPC_URL, headers=headers, json=payload, timeout=10)
        result = response.json().get("result")

        if not result or 'tokenTransfers' not in result:
            print("⚠️ No token transfers, skipping.")
            continue

        for transfer in result['tokenTransfers']:
            token_transfers.append({
                "signature": sig,
                "source": transfer.get("source"),
                "destination": transfer.get("destination"),
                "mint": transfer.get("mint"),
                "amount": transfer.get("amount"),
                "decimals": transfer.get("decimals"),
                "tokenName": transfer.get("tokenName"),
                "tokenSymbol": transfer.get("tokenSymbol"),
                "timestamp": result.get("blockTime"),
            })

        sleep(0.1)
    except Exception as e:
        print(f"❌ Error processing {sig}: {e}")
        continue

# === Save token transfers ===
with open("token_transfers.json", "w") as tf_out:
    json.dump(token_transfers, tf_out, indent=2)

print(f"✅ Saved {len(token_transfers)} token transfers to token_transfers.json")

File: solana-monitoring/fetch/get_token_transfers.js
#!/usr/bin/env node

const fs = require('fs');
const fetch = require('node-fetch');
const minimist = require('minimist');

const args = minimist(process.argv.slice(2));
const signatureFile = args._[0];
const outputFile = args.output || 'direction_counts_by_token_transfer_type.json';
const RPC_URL = 'https://devnet.helius-rpc.com/?api-key=934c08b5-0c44-4549-bda8-30587b901642';

if (!signatureFile) {
  console.error('Usage: node get_token_transfers.js <signature_file> [--output output_file]');
  process.exit(1);
}

// === Load and parse JSON signatures ===
const rawData = JSON.parse(fs.readFileSync(signatureFile, 'utf8'));
let signatures = [];
for (const entry of rawData) {
  const result = entry.result || {};
  const sigs = result.signatures || [];
  signatures = signatures.concat(sigs);
}
console.log(`\uD83D\uDCE6 Loaded ${signatures.length} signatures.`);

const tokenTransfers = [];

(async () => {
  for (let i = 0; i < signatures.length; i++) {
    const sig = signatures[i];
    try {
      console.log(`[${i + 1}/${signatures.length}] Fetching token transfers for ${sig}...`);
      const payload = {
        jsonrpc: '2.0',
        id: 1,
        method: 'getTokenTransfers',
        params: [sig, 'jsonParsed']
      };
      const res = await fetch(RPC_URL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
        timeout: 10000
      });
      const data = await res.json();
      const result = data.result;
      if (!result || !result.tokenTransfers) {
        console.log('⚠️ No token transfers, skipping.');
        continue;
      }
      for (const transfer of result.tokenTransfers) {
        tokenTransfers.push({
          signature: sig,
          source: transfer.source,
          destination: transfer.destination,
          mint: transfer.mint,
          amount: transfer.amount,
          decimals: transfer.decimals,
          tokenName: transfer.tokenName,
          tokenSymbol: transfer.tokenSymbol,
          timestamp: result.blockTime
        });
      }
      await new Promise(r => setTimeout(r, 100));
    } catch (e) {
      console.log(`❌ Error processing ${sig}: ${e}`);
      continue;
    }
  }
  fs.writeFileSync('token_transfers.json', JSON.stringify(tokenTransfers, null, 2));
  console.log(`✅ Saved ${tokenTransfers.length} token transfers to token_transfers.json`);
})(); 